central/hash/manager/deduper_test.go

```diff
@@ -321,3 +322,195 @@ func TestReconciliation(t *testing.T) {
+				syncEventShouldNotProcess(d1),
+				// d2 should be processed
+				syncEventSuccessfully(d2),
+				// Sync event
+				syncEvent,
+				// Assert d1 and d2
+				assertEvents([]*central.MsgFromSensor{d1, d2}),
+			},
+		},
+		"reconnection (sensor can handle the deduper state)": {
+			events: []testEvents{
+				// Simulated new connection
+				newConnection(nil),
+				// Sync resources d1 d2
+				syncEventSuccessfully(d1),
+				syncEventSuccessfully(d2),
+				// Sync event
+				syncEvent,
+				// Assert d1 and d2
+				assertEvents([]*central.MsgFromSensor{d1, d2}),
+				// Simulated reconnection
+				newConnection(nil),
+				// Sensor does not send the sync resources d1 d2 as it can handle the deduper state
+				syncEventSuccessfully(d3),
+				// Sync event
+				syncEvent,
+				// Assert d1, d2, and d3
+				assertEvents([]*central.MsgFromSensor{ /* d1, d2, */ d3}), // FIXME: we should have d1 and d2
+			},
+		},
+		"reconnection with unsuccessful events (sensor can handle the deduper state)": {
+			events: []testEvents{
+				// Simulated new connection
+				newConnection(nil),
+				// Sync resources d1 d2
+				syncEventSuccessfully(d1),
+				syncEventUnsuccessfully(d2),
+				// Simulated reconnection
+				newConnection(nil),
+				// Sensor does not send the sync resources d1 as it can handle the deduper state
+				syncEventSuccessfully(d2),
+				syncEventSuccessfully(d3),
+				// Sync event
+				syncEvent,
+				// Assert d1, d2, and d3
+				assertEvents([]*central.MsgFromSensor{ /* d1, */ d2, d3}),
+			},
+		},
+	}
+	for tname, tc := range cases {
+		t.Run(tname, func(tt *testing.T) {
+			var deduper *deduperImpl
+			for _, event := range tc.events {
+				event(tt, &deduper)
+			}
+		})
+	}
+}
+
+func newConnection(initialHashes map[string]uint64) testEvents {
+	return func(_ *testing.T, deduper **deduperImpl) {
+		if *deduper == nil {
+			*deduper = NewDeduper(initialHashes).(*deduperImpl)
+		}
+		(*deduper).StartSync()
+	}
+}
+
+func syncEventSuccessfully(event *central.MsgFromSensor) testEvents {
+	return func(t *testing.T, deduper **deduperImpl) {
+		assert.True(t, (*deduper).ShouldProcess(event))
+		(*deduper).MarkSuccessful(event)
+	}
+}
+
+func syncEventShouldNotProcess(event *central.MsgFromSensor) testEvents {
+	return func(t *testing.T, deduper **deduperImpl) {
+		assert.False(t, (*deduper).ShouldProcess(event))
+	}
+}
+
+func syncEventUnsuccessfully(event *central.MsgFromSensor) testEvents {
+	return func(t *testing.T, deduper **deduperImpl) {
+		assert.True(t, (*deduper).ShouldProcess(event))
+	}
+}
+
+func syncEvent(_ *testing.T, deduper **deduperImpl) {
+	(*deduper).ProcessSync()
+}
+
+func assertEvents(events []*central.MsgFromSensor) testEvents {
+	return func(t *testing.T, deduper **deduperImpl) {
+		assert.Len(t, (*deduper).successfullyProcessed, len(events))
+		for _, event := range events {
+			_, ok := (*deduper).successfullyProcessed[eventPkg.GetKeyFromMessage(event)]
```

Comment:
nitpick: (for future modifications) we could use `assert.Contains`.

Reply:
Yeah let's do this in a follow-up to not wait for CI